tortoise/verifying_tortoise.go

```diff
@@ -0,0 +1,574 @@
+	// TODO: fix potential leak when we can't verify but keep receiving layers
+
+	if t.Verified <= types.GetEffectiveGenesis()+t.Hdist {
+		return
+	}
+	// The window is the last [Verified - hdist] layers.
+	window := t.Verified - t.Hdist
+	t.logger.Info("Window starts %v", window)
+	// evict from last evicted to the beginning of our window.
+	for lyr := t.Evict; lyr < window; lyr++ {
+		t.logger.Info("removing lyr %v", lyr)
+		for blk := range t.BlocksToBlocks[lyr] {
+			delete(t.GoodBlocksIndex, blk)
+		}
+		delete(t.BlocksToBlocks, lyr)
+		t.logger.Debug("evict block %v from maps ", lyr)
+
+	}
+	t.Evict = window
+}
+
+func removeOpinion(o []Opinion, i int) []Opinion {
+	return append(o[:i], o[i+1:]...)
+}
+
+// TODO: cache but somehow check for changes (hare that didn't finish finishes..) maybe check hash?
+func (t *turtle) getSingleInputVectorFromDB(lyrid types.LayerID, blockid types.BlockID) (vec, error) {
+	if lyrid <= types.GetEffectiveGenesis() {
+		return support, nil
+	}
+
+	input, err := t.bdp.GetLayerInputVector(lyrid)
+	if err != nil {
+		return abstain, err
+	}
+
+	str := "["
+	for i, b := range input {
+		str += b.String()
+		if i != len(input)-1 {
+			str += ","
+		}
+	}
+	str += "]"
+	t.logger.With().Debug("got input vector from db", lyrid, log.FieldNamed("query_block", blockid), log.String("input", str))
+
+	for _, bl := range input {
+		if bl == blockid {
+			return support, nil
+		}
+	}
+
+	return against, nil
+}
+
+func (t *turtle) inputVectorForLayer(lyrBlocks []types.BlockID, input []types.BlockID) map[types.BlockID]vec {
+	lyrResult := make(map[types.BlockID]vec, len(lyrBlocks))
+	//XXX : input vector must be pointer so we can differentiate
+	// no support votes to no results at all.
+	if input == nil {
+		// hare didn't finish hence we don't have Opinion
+		// TODO: get hare Opinion when hare finishes
+		for _, b := range lyrBlocks {
+			lyrResult[b] = abstain
+		}
+		return lyrResult
+	}
+
+	for _, b := range input {
+		lyrResult[b] = support
+	}
+
+	for _, b := range lyrBlocks {
+		if _, ok := lyrResult[b]; !ok {
+			lyrResult[b] = against
+		}
+	}
+	return lyrResult
+}
+
+func (t *turtle) BaseBlock() (types.BlockID, [][]types.BlockID, error) {
+	// Try to find a block counting only good blocks
+	for i := t.Last; i > t.Last-t.Hdist; i-- {
+		for blk, op := range t.BlocksToBlocks[i] {
+			if _, ok := t.GoodBlocksIndex[blk]; !ok {
+				t.logger.With().Debug("can't use a not good block", log.FieldNamed("last_layer", t.Last), i, blk)
+				continue
+			}
+			afn, err := t.opinionMatches(i, blk, op)
+			if err != nil {
+				t.logger.With().Debug("can't use block error %v", log.FieldNamed("last_layer", t.Last), i, blk, log.Err(err))
+				continue
+			}
+
+			t.logger.With().Info("Choose baseblock", log.FieldNamed("last_layer", t.Last), log.FieldNamed("base_block_layer", i), blk, log.Int("against_count", len(afn[0])), log.Int("support_count", len(afn[1])), log.Int("neutral_count", len(afn[2])))
```

Comment:
remove capital letter

Reply:
why? we shouldn't use capital letters when logging?

Reply:
no, we should have one standard which is non capital letters